Lib/Inc/CPIDLoop.h

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file CPIDLoop.h
+ */
+
+/*
+ * CPIDLoop.h
+ *
+ *  Created on: 19 Feb 2022
+ *      Author: salavat.magazov
+ */
+
+#ifndef CPIDLOOP_H_
+#define CPIDLOOP_H_
+
+/* Default PID parameters.
+ * These are setup to make PID loop transparent. Output will be equal to error.
+ * Proportional is percentage of power per unit of error.
+ * Integral is units of output per unit of error per unit of time.
+ * Differential is unit of output per degree error change per unit of time.
+ * Normally for slow moving processes differential is zero.
+ */
+#define DEFAULT_P_COEFF 1
+#define DEFAULT_I_COEFF 0
+#define DEFAULT_D_COEFF 0
+
+class CPIDLoop
+{
+public:
+    struct CONTROL_STRUCT_T
+    {
+        CONTROL_STRUCT_T(float k_p, float k_i, float k_d);
+
+        float p_coeff;
+        float i_coeff;
+        float d_coeff;
+
+        float min_out;
+        float max_out;
+
+        float min_i;
+        float max_i;
+
+        float min_d;
+        float max_d;
+    };
+
+    CPIDLoop(float k_p = DEFAULT_P_COEFF,
+             float k_i = DEFAULT_I_COEFF,
+             float k_d = DEFAULT_D_COEFF);
+
+    float run(float target, float actual);
+    float run(float target, float actual, float deltaError);
+    void reset();
+
+    /* PID runtime terms access methods. */
+    float getP() const
+    {
+        return m_p;
+    }
+
+    float getI() const
+    {
+        return m_i;
+    }
+    float getD() const
+    {
+        return m_d;
+    }
+    void setI(float i)
```

Comment:
Because both P and D are calculated based on instantaneous input. Integral on the other hand has "memory" of past error values. If you want to modify that "memory" you need to access the value.